stratosphere/pm/source/impl/pm_process_manager.cpp

```diff
@@ -0,0 +1,680 @@
+
+            /* Set flags. */
+            if (is_application) {
+                process_info->SetApplication();
+            }
+            if (ShouldSignalOnStart(args->flags) && allow_debug) {
+                process_info->SetSignalOnStart();
+            }
+            if (ShouldSignalOnExit(args->flags)) {
+                process_info->SetSignalOnExit();
+            }
+            if (ShouldSignalOnDebugEvent(args->flags) && allow_debug) {
+                process_info->SetSignalOnDebugEvent();
+            }
+
+            /* Process hooks/signaling. */
+            if (location.title_id == g_title_id_hook) {
+                g_hook_to_create_process_event->Signal();
+                g_title_id_hook = ncm::InvalidTitleId;
+            } else if (is_application && g_application_hook) {
+                g_hook_to_create_application_process_event->Signal();
+                g_application_hook = false;
+            } else if (!ShouldStartSuspended(args->flags)) {
+                R_TRY(StartProcess(process_info, &program_info));
+            }
+
+            /* Add process to list. */
+            {
+                ProcessListAccessor list(g_process_list);
+                list->Add(process_info);
+                g_process_waitable_manager->AddWaitable(new ProcessInfoWaiter(process_info));
+            }
+
+            *args->out_process_id = process_id;
+            return ResultSuccess;
+        }
+
+        Result LaunchProcessEventCallback(u64 timeout) {
+            g_process_launch_start_event->Clear();
+            g_process_launch_result = LaunchProcess(&g_process_launch_args);
+            g_process_launch_finish_signal.Signal();
+            return ResultSuccess;
+        }
+
+        void CleanupProcess(ProcessListAccessor &list, std::shared_ptr<ProcessInfo> process_info) {
+            /* Remove the process from the list. */
+            list->Remove(process_info->GetProcessId());
+
+            /* Close process resources. */
+            process_info->Cleanup();
+
+            /* Handle the case where we need to keep the process alive some time longer. */
+            if (GetRuntimeFirmwareVersion() >= FirmwareVersion_500 && process_info->ShouldSignalOnExit()) {
+                /* Add the process to the list of dead processes. */
+                {
+                    ProcessListAccessor dead_list(g_dead_process_list);
+                    dead_list->Add(process_info);
+                }
+                /* Signal. */
+                g_process_event->Signal();
+            }
+        }
+
+    }
+
+    /* Initialization. */
+    Result InitializeProcessManager() {
+        /* Create events. */
+        g_process_event = CreateWriteOnlySystemEvent();
+        g_hook_to_create_process_event = CreateWriteOnlySystemEvent();
+        g_hook_to_create_application_process_event = CreateWriteOnlySystemEvent();
+        g_boot_finished_event = CreateWriteOnlySystemEvent();
```

Comment:
Transferring ownership by moving pointers around isn't very C++, but that's more of a general stratosphere concern than a PM concern.

Reply:
Yeah, problem as always is that I can't statically initialize these :/

It's kind of a design issue. Is there any way to make __appInit happen before static initializers are called?

Reply:
Pinging @fincs

Reply:
[__appInit already happens before static initializers](https://github.com/switchbrew/libnx/blob/master/nx/source/runtime/init.c#L177-L182)

Reply:
Okay, cool. I'll statically initialize these then.